Set up isolated PATH for completion test suites

FloGa opened the pull request:
As discussed in #1923, the test cases will fail if there is a command in PATH which starts with "true". This PR fixes this problem by creating a temporary directory, symlinking needed commands, and using this directory as the only item in the PATH variable.

blueyed commented on spec/awful/completion_spec.lua:
```diff
@@ -15,7 +15,38 @@ if not has_bash and not has_zsh then
+        local target = GLib.find_program_in_path(cmd)
+        if target then
+            os.execute(string.format('ln -s %s %s/%s', target, test_path, cmd))
+        end
```
Could have an error in case it is not found?

blueyed replied:
Is there something in GLib to create the symlink maybe?

FloGa replied:
According to https://developer.gnome.org/glib/unstable/glib-Miscellaneous-Utility-Functions.html#g-find-program-in-path, if file is not found, NULL is returned, hence the `if target`. Since you also check in every test if bash/zsh is available, a missing file would not cause an error.

Regarding the symlink, I now triple checked the API and haven't found a `link` method or anything that resembles this functionality ... But it has a `chmod` method. Should I stick to my original plan and create executable files that call the appropriate command?

    local link = io.open(test_path .. "/" .. cmd, "w")
    link:write("exec " .. target .. ' "$@"')
    link:close()
    GLib.chmod(test_path .. "/" .. cmd, 0700)

(Not sure about the octal representation of the file mode, I would test it if you want to go with this solution.)

blueyed replied:
IMHO the `ln -s` is fine then.

psychon replied:
https://developer.gnome.org/gio/stable/GFile.html#g-file-make-symbolic-link

psychon replied:
```
lua -e 'print(require("lgi").Gio.File.new_for_path("/tmp/foo"):make_symbolic_link("target", nil))'
```

FloGa replied:
Nice, I didn't see this one. I guess I'm going with this then.

blueyed commented on spec/awful/completion_spec.lua:
```diff
@@ -15,7 +15,38 @@ if not has_bash and not has_zsh then
     return
 end
 
+local orig_path = GLib.getenv("PATH")
```
Name this `orig_PATH` maybe for clarity, or `orig_path_env`?

FloGa replied:
I would go with `orig_path_env` to keep it lowercase.

blueyed replied:
OK.

FloGa commented:
I changed the variable name and replaced the os call with a Gio method.

As before, to keep the discussion history I just did fixup commits. If you approve the PR, I will rebase it onto the current master and resolve all fixups.

FloGa commented:
Thank you for approval. I rebased on current master and squashed the fixup commits.

psychon commented on spec/awful/completion_spec.lua:
```diff
@@ -15,6 +15,37 @@ if not has_bash and not has_zsh then
+        local target = GLib.find_program_in_path(cmd)
+        if target then
+            local link = string.format('%s/%s', test_path, cmd)
+            Gio.File.new_for_path(link):make_symbolic_link(target)
```
It does not really make a difference, but:
```lua
local test_path = Gio.file.new_tmp("awesome-tests-path-XXXXXX")
test_path:delete()
test_path:make_directory()

for _, cmd in ipairs(required_commands) do
    local target = GLib.find_program_in_path(cmd)
    if target then
        target:get_child(cmd):make_symbolic_link(target)
    end
```

FloGa replied:
`target` is a string, not a GFile object, hence there is no `get_child()`. Shall I instantiate `target` as a new GFile object, then?

psychon replied:
Whoops, my fault. Let's try it step-by-step this time:
```lua
local link = string.format('%s/%s', test_path, cmd)
Gio.File.new_for_path(link):make_symbolic_link(target)
```
This constructs `cmd` as a child of `test_path` and then creates a GFile instance for it. So what I should have written is `test_path:get_child(cmd):make_symbolic_link(target)`. I'll update my above comment for this.

FloGa replied:
Ah, now I see. I will keep returning the path as a string however, since it is needed in other parts for easier deletion of this whole directory (like in the existing `get_test_dir` method).

FloGa commented:
Added the requested changes to simplify creation of testing path environment and symlinking needed commands.

blueyed commented:
Thank you!
